draft-thomson-http-oblivious.md

```diff
@@ -381,6 +381,226 @@ Plaintext Messages support arbitrary length padding. Clients and servers MAY pad
+POST request to the oblivious proxy resource.  This request MUST only contain
+those fields necessary to carry the encapsulated request: a method of POST, a
+target URI of the URI of the oblivious proxy resource, a header field containing
+the content type (see ({{media-types}}), and the encapsulated request as the
+request content.  Clients MAY include fields that do not reveal information
+about the content of the request, such as Alt-Used {{?ALT-SVC=RFC7838}}.
+
+The oblivious proxy resource interacts with the oblivious request resource by
+constructing a request using the same restrictions as the client request, except
+that the target URI is the oblivious request resource.  The content of this
+request is copied from the client.  The oblivious proxy resource MUST NOT add
+information about the client to this request.
+
+When a response is received from the oblivious request resource, the oblivious
+proxy resource forwards the response according to the rules of an HTTP proxy;
+see Section 7.6 of {{!HTTP}}.
+
+An oblivious request resource, if it receives any response from the oblivious
+target resource, sends a single 200 response containing the encapsulated
+response.  Like the request from the client, this response MUST only contain
+those fields necessary to carry the encapsulated response: a 200 status code, a
+header field indicating the content type, and the encapsulated response as the
+response content.  As with requests, additional fields MAY be used to convey
+information that does not reveal information about the encapsulated response.
+
+An oblivious request resource acts as a gateway for requests to the oblivious
+target resource (see Section 7.6 of {{!HTTP}}).  The one exception is that any
+information it might forward in a response MUST be encapsulated, unless it is
+responding to errors it detects before removing encapsulation of the request;
+see {{errors}}.
+
+
+## Informational Responses
+
+This encapsulation does not permit progressive processing of responses.  Though
+the binary HTTP response format does support the inclusion of informational
+(1xx) status codes, the AEAD encapsulation cannot be removed until the entire
+message is received.
+
+In particular, the Expect header field with 100-continue (see Section 10.1.1 of
+{{!HTTP=I-D.ietf-httpbis-semantics}}) cannot be used.  Clients MUST NOT
+construct a request that includes a 100-continue expectation; the oblivious
+request resource MUST generate an error if a 100-continue expectation is
+received.
+
+
+## Errors
+
+A server that receives an invalid message for any reason MUST generate an HTTP
+response with a 4xx status code.
+
+Errors detected by the oblivious proxy resource and errors detected by the
+oblivious request resource before removing protection (including being unable to
+remove encapsulation for any reason) result in the status code being sent
+without protection in response to the POST request made to that resource.
```

Comment:
```suggestion
without protection in response to the POST request made to that resource.
Oblivious request resources MUST use distinct status codes to distinguish 
HPKE key mismatch from other errors, such as decapsulation failure. This is
needed to inform clients of stale or otherwise incorrect HPKE keying material.
```

Reply:
Feel free to include this in some other way, but I think it's important to note. Without a clear "your key is out of date" signal, clients don't know if and when they should fetch a fresh set of keys for encapsulation.

Reply:
Let's open an issue for this.  I think we want to recommend a status code for this, but it is possible that there isn't something easy we can use here (other than to maybe suggest that clients re-fetch key configuration when they get a non-encapsulated error response).